File: app/src/main/java/me/ykrank/s1next/view/adapter/EmoticonGridRecyclerAdapter.java
package me.ykrank.s1next.view.adapter;

import android.app.Activity;
import android.databinding.DataBindingUtil;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import com.bumptech.glide.Glide;
import com.bumptech.glide.RequestManager;

import java.util.List;

import me.ykrank.s1next.App;
import me.ykrank.s1next.R;
import me.ykrank.s1next.data.api.model.Emoticon;
import me.ykrank.s1next.databinding.ItemEmoticonBinding;
import me.ykrank.s1next.viewmodel.EmoticonViewModel;
import me.ykrank.s1next.widget.RxBus;

public final class EmoticonGridRecyclerAdapter
        extends RecyclerView.Adapter<EmoticonGridRecyclerAdapter.BindingViewHolder> {

    private final LayoutInflater mLayoutInflater;

    private final List<Emoticon> mEmoticons;
    private final RequestManager mEmoticonRequestBuilder;

    private final RxBus mRxBus;

    public EmoticonGridRecyclerAdapter(Activity activity, List<Emoticon> emoticons) {
        mLayoutInflater = activity.getLayoutInflater();
        this.mEmoticons = emoticons;
        mEmoticonRequestBuilder = Glide.with(activity);
        mRxBus = App.getAppComponent().getEventBus();

        setHasStableIds(true);
    }

    @Override
    public BindingViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        ItemEmoticonBinding binding = DataBindingUtil.inflate(mLayoutInflater,
                R.layout.item_emoticon, parent, false);
        binding.setEventBus(mRxBus);
        binding.setRequestManager(mEmoticonRequestBuilder);
        binding.setEmoticonViewModel(new EmoticonViewModel());

        return new BindingViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(BindingViewHolder holder, int position) {
        holder.itemEmoticonBinding.getEmoticonViewModel().emoticon.set(mEmoticons.get(position));
        holder.itemEmoticonBinding.executePendingBindings();
    }

    @Override
    public void onViewRecycled(BindingViewHolder holder) {
        super.onViewRecycled(holder);
    }

    @Override
    public int getItemCount() {
        return mEmoticons.size();
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    public static final class BindingViewHolder extends RecyclerView.ViewHolder {

        private final ItemEmoticonBinding itemEmoticonBinding;

        public BindingViewHolder(ItemEmoticonBinding itemEmoticonBinding) {
            super(itemEmoticonBinding.getRoot());

            this.itemEmoticonBinding = itemEmoticonBinding;
        }
    }
}

File: app/src/main/java/me/ykrank/s1next/view/adapter/EmoticonGridRecyclerAdapter.kt
package me.ykrank.s1next.view.adapter

import android.app.Activity
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup

import com.bumptech.glide.Glide
import com.bumptech.glide.RequestManager
import com.github.ykrank.androidtools.widget.RxBus

import me.ykrank.s1next.App
import me.ykrank.s1next.R
import me.ykrank.s1next.data.api.model.Emoticon
import me.ykrank.s1next.databinding.ItemEmoticonBinding
import me.ykrank.s1next.viewmodel.EmoticonViewModel

class EmoticonGridRecyclerAdapter(activity: Activity, private val mEmoticons: List<Emoticon>) : androidx.recyclerview.widget.RecyclerView.Adapter<EmoticonGridRecyclerAdapter.BindingViewHolder>() {

    private val mLayoutInflater: LayoutInflater = activity.layoutInflater
    private val mEmoticonRequestBuilder: RequestManager = Glide.with(activity)

    private val mRxBus: RxBus = App.preAppComponent.rxBus

    init {
        setHasStableIds(true)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BindingViewHolder {
        val binding = DataBindingUtil.inflate<ItemEmoticonBinding>(mLayoutInflater,
                R.layout.item_emoticon, parent, false)
        binding.rxBus = mRxBus
        binding.requestManager = mEmoticonRequestBuilder
        binding.emoticonViewModel = EmoticonViewModel()

        return BindingViewHolder(binding)
    }

    override fun onBindViewHolder(holder: BindingViewHolder, position: Int) {
        holder.itemEmoticonBinding.emoticonViewModel?.emoticon?.set(mEmoticons[position])
    }

    override fun onViewRecycled(holder: BindingViewHolder) {
        super.onViewRecycled(holder)
    }

    override fun getItemCount(): Int {
        return mEmoticons.size
    }

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    class BindingViewHolder(val itemEmoticonBinding: ItemEmoticonBinding) : androidx.recyclerview.widget.RecyclerView.ViewHolder(itemEmoticonBinding.root)
}
